dask/distributed/tests/test_scheduler.py

```diff
@@ -229,3 +233,100 @@ def test_close_scheduler():
     assert not s._listen_to_clients_thread.is_alive()
     assert not s._listen_to_workers_thread.is_alive()
     assert s.context.closed
+
+
+def test_prune_and_notify():
+    with scheduler_and_workers(worker_kwargs={'heartbeat': 0.001}) as (s, (w1, w2)):
+        # Oh no! A worker died!
+        w2.close()
+        while w2.status != 'closed':
+            sleep(1e-6)
+
+        # worker 1 tries to collect data from the dead worker.
+        result = w1.pool.apply_async(w1.collect, args=({'x': [w2.address]},))
+        sleep(0.01)  # sleep to show w1.collect hangs
+        assert result.ready() is False
+
+        # The scheduler notices, and corrects it state.
+        while w2.address in s.workers:
+            s.prune_and_notify(timeout=0.01)
+
+        # But the sheduler notified the workers about the death
+        while not result.ready():
+            sleep(1e-6)
+        assert result.ready() is True
+
+
+def test_workers_reregister():
+    with scheduler_and_workers(worker_kwargs={'heartbeat': 0.001}) as (s, (w1, w2)):
+        assert w1.address in s.workers
+
+        assert s.workers.pop(w1.address)
+        assert w1.address not in s.workers  # removed
+
+        while w1.address not in s.workers:
+            sleep(1e-6)
+        assert w1.address in s.workers  # but now its back
+
+
+def test_collect_retry():
+    with scheduler_and_workers(n=3, worker_kwargs={'heartbeat': 0.001}) as (s, (w1, w2, w3)):
+        w3.data['x'] = 42
+        w2.close()
+        while w2.status != 'closed':  # make sure closed
+            sleep(1e-6)
+
+        result = w1.pool.apply_async(w1.collect,
+                                     args=({'x': [w2.address, w3.address]},))
+        while w2.address in s.workers:
+            s.prune_and_notify(timeout=0.01)
+
+        # make sure prune_and_notify didn't remove these
+        assert w1.address in s.workers
+        assert w3.address in s.workers
+
+        while not result.ready():  # make sure collect finishes
+            sleep(1e-6)
+        assert result.ready() == True
+        assert 'x' in w1.data
+        assert w1.data['x'] == 42
+
+
+def test_collect():
+    with scheduler_and_workers() as (s, (w1, w2)):
+        w1.data['x'] = 42
+        w2.collect({'x': [w1.address]})
+        assert w2.data['x'] == 42
+
+
+def test_worker_death():
```

Comment:
This and the test above don't actually use a scheduler. So I'll move it to test_workers.py